kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java

```diff
@@ -0,0 +1,173 @@
+        })))
+                .thenApply(HttpResponse::body)
+                .thenApply(Supplier::get)
+                .thenApply(VaultResponse::data)
+                .thenApply(data -> {
+                    var secretKey = new SecretKeySpec(Base64.getDecoder().decode(data.plaintext()), AES_KEY_ALGO);
+                    return new DekPair<>(new VaultEdek(kekRef, data.ciphertext().getBytes(StandardCharsets.UTF_8)), secretKey);
+                });
+
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * @see <a href="https://developer.hashicorp.com/vault/api-docs/secret/transit#decrypt">https://developer.hashicorp.com/vault/api-docs/secret/transit#decrypt</a>
+     */
+    @NonNull
+    @Override
+    public CompletionStage<SecretKey> decryptEdek(@NonNull VaultEdek edek) {
+
+        var body = createDecryptPostBody(edek);
+
+        var request = createVaultRequest()
+                .uri(vaultUrl.resolve("v1/transit/decrypt/%s".formatted(edek.kekRef())))
+                .POST(HttpRequest.BodyPublishers.ofString(body))
+                .build();
+
+        return vaultClient.sendAsync(request, statusHandler(edek.kekRef(), new JsonBodyHandler<VaultResponse<DecryptData>>(new TypeReference<>() {
+        }))).thenApply(HttpResponse::body)
+                .thenApply(Supplier::get)
+                .thenApply(VaultResponse::data)
+                .thenApply(data -> new SecretKeySpec(Base64.getDecoder().decode(data.plaintext()), AES_KEY_ALGO));
+    }
+
+    private String createDecryptPostBody(@NonNull VaultEdek edek) {
+        var map = Map.of("ciphertext", new String(edek.edek(), StandardCharsets.UTF_8));
+
+        try {
+            return new ObjectMapper().writeValueAsString(map);
+        }
+        catch (JsonProcessingException e) {
+            throw new KmsException("Failed to build request body for %s".formatted(edek.kekRef()));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * @see <a href="https://developer.hashicorp.com/vault/api-docs/secret/transit#read-key">https://developer.hashicorp.com/vault/api-docs/secret/transit#read-key</a>
+     */
+    @NonNull
+    @Override
+    public CompletableFuture<String> resolveAlias(@NonNull String alias) {
+
+        var request = createVaultRequest()
+                .uri(vaultUrl.resolve("v1/transit/keys/%s".formatted(alias)))
+                .build();
+
+        return vaultClient.sendAsync(request, statusHandler(alias, new JsonBodyHandler<VaultResponse<ReadKeyData>>(new TypeReference<>() {
+        })))
+                .thenApply(HttpResponse::body)
+                .thenApply(Supplier::get)
+                .thenApply(VaultResponse::data)
+                .thenApply(ReadKeyData::name);
+
+    }
+
+    @NonNull
+    @Override
+    public Serde<VaultEdek> edekSerde() {
+        return new VaultEdekSerde();
+    }
+
+    private HttpRequest.Builder createVaultRequest() {
+        return HttpRequest.newBuilder()
```

Comment:
We could set a hardcoded [timeout](https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/HttpRequest.Builder.html#timeout(java.time.Duration)) here to match the connect timeout on the client.

> Sets a timeout for this request. If the response is not received within the specified timeout then an [HttpTimeoutException](https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/HttpTimeoutException.html) is thrown from [HttpClient::send](https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/HttpClient.html#send(java.net.http.HttpRequest,java.net.http.HttpResponse.BodyHandler)) or [HttpClient::sendAsync](https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/HttpClient.html#sendAsync(java.net.http.HttpRequest,java.net.http.HttpResponse.BodyHandler)) completes exceptionally with an HttpTimeoutException. The effect of not setting a timeout is the same as setting an infinite Duration, i.e. block forever.

Though it looks like there's some problems with it in that the request timeout is cancelled when the headers are received, so a slow/blocked response body can still block it up forever[JDK-8258397]( https://bugs.openjdk.org/browse/JDK-8258397). The suggested workaround we might want to do later is ensure we set up a scheduled timeout for our own Future and cancel the request.